<a href="https://colab.research.google.com/github/Stollere/test_statistik/blob/main/KIundBigData_GruppeK.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Datensatz

The dataset we are going to use for this purpose was originally published by the World Health Organization (WHO).

Regression - the prediction of a continuous target variable

# Datensatz laden

In [6]:
!gdown https://drive.google.com/uc?id=10Rg1GjBIqP4rcpLMjIUj2_64YREsDXGf

Downloading...
From: https://drive.google.com/uc?id=10Rg1GjBIqP4rcpLMjIUj2_64YREsDXGf
To: /content/life_expectancy.csv
100% 333k/333k [00:00<00:00, 108MB/s]


In [7]:
import pandas as pd

In [8]:
life_expectancy = pd.read_csv('life_expectancy.csv')

# Datenexploration

In [5]:
life_expectancy

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
0     ...    6.0               8.16         65.0        0.1  584.259210   
1     ...   58.0               8.18         62.0        0.1  612.696514   
2     ...   62.0               8.13         64.0        0.1  631.744976   
3     ...   67.0               8.52         67.0        0.1  669.959000   
4     ...   68.0               7.87         68.0        0.1   63.537231   
...   ...    ...                ...          ...        ...         ...   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
1       327582.0                   17.5                 17.5   
2     31731688.0                   17.7                 17.7   
3      3696958.0                   17.9                 18.0   
4      2978599.0                   18.2                 18.2   
...          ...                    ...                  ...   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427      

In [9]:
life_expectancy['Life expectancy'].mean()

KeyError: ignored

# Data Cleansing

In [10]:
life_expectancy.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

Some of the dataset columns have whitespace characters at the beginning or end so we can use the strip method to remove those:

In [11]:
life_expectancy = life_expectancy.rename(str.strip, axis='columns')

In [12]:
life_expectancy.columns

Index(['Country', 'Year', 'Status', 'Life expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles', 'BMI', 'under-five deaths', 'Polio', 'Total expenditure',
       'Diphtheria', 'HIV/AIDS', 'GDP', 'Population', 'thinness  1-19 years',
       'thinness 5-9 years', 'Income composition of resources', 'Schooling'],
      dtype='object')

We will begin by exploring and prepocessing our dataset.
Let's start by sampling 15 random rows:

In [14]:
life_expectancy.sample(15)

Country  Year      Status  Life expectancy  Adult Mortality  \
678               Cyprus  2010   Developed             79.5             59.0   
2348            Slovenia  2012   Developed             82.0             81.0   
2535         Switzerland  2001   Developed             82.0             75.0   
2435               Spain  2005   Developed             81.0             77.0   
1828               Nepal  2000  Developing             62.5            238.0   
775   Dominican Republic  2010  Developing             72.7            166.0   
2023         Philippines  2015  Developing             68.5            211.0   
2934            Zimbabwe  2003  Developing             44.5            715.0   
1234                Iraq  2015  Developing             68.9            182.0   
2309        Sierra Leone  2003  Developing             41.5             57.0   
1865           Nicaragua  2011  Developing             74.5            147.0   
462           Cabo Verde  2001  Developing             73.0            152.0   
1167             Hungary  2002   Developed             72.5            184.0   
2243        Saudi Arabia  2005  Developing             73.1             11.0   
1399              Kuwait  2010  Developing             74.0             86.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles  \
678               0    11.32              283.330026         96.0       18   
2348              0    10.95             3283.024864          NaN        2   
2535              0    11.12             6478.346135          NaN      700   
2435              2    11.92              405.447598         96.0       22   
1828             46     0.08               17.912337          NaN     9397   
775               6     6.02              664.558524         83.0        0   
2023             52      NaN                0.000000          6.0      619   
2934             26     4.06                0.000000          7.0      998   
1234             32      NaN                0.000000         56.0     1433   
2309             30     4.07               38.614732          NaN      586   
1865              3     3.39              321.613259         98.0        0   
462               0     3.81              150.743486          NaN        0   
1167              1    13.29                8.986198          NaN        0   
2243              9     0.05             1117.048134         97.0      373   
1399              1     0.01             2009.575605         99.0       13   

      ...  Polio  Total expenditure  Diphtheria  HIV/AIDS           GDP  \
678   ...   99.0               7.23        99.0       0.1   3818.463960   
2348  ...   96.0               9.37        96.0       0.1  22486.471670   
2535  ...   95.0               1.28        93.0       0.1  38538.644470   
2435  ...   96.0               8.12        96.0       0.1   2651.717450   
1828  ...   74.0               5.43        74.0       0.1    231.425540   
775   ...   85.0               4.60        88.0       0.9   5451.669600   
2023  ...   79.0                NaN         6.0       0.1   2878.338370   
2934  ...    7.0               6.52        68.0      36.7    453.351155   
1234  ...   63.0                NaN        58.0       0.1   4974.269300   
2309  ...   66.0              11.69        73.0       1.9    263.761831   
1865  ...   99.0               6.39        98.0       0.1   1682.957924   
462   ...   91.0               5.19         9.0       0.8   1268.884564   
1167  ...   98.0               7.47        99.0       0.1    665.644300   
2243  ...   96.0               3.42        96.0       0.1  13739.829450   
1399  ...   98.0               2.76        98.0       0.1  38497.616960   

      Population  thinness  1-19 years  thinness 5-9 years  \
678     111267.0                   0.9                 1.0   
2348    257159.0                   1.5                 1.5   
2535   7229854.0                   0.7                 0.5   
2435  43653155.0                   0.6

Each row in the dataset describes a country in a given year. Independent variables include various attributes on health, education, economic status and population. The dependent variable of this dataset is the life expectancy in years. Our goal will be to build a model to predict the life expectancy of any given country.

If we take a look at the types we can see that we are mostly dealing with numerical attributes so we won't have to do much preprocessing in terms of encoding.

In [15]:
life_expectancy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10  BMI                              2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

Actually, the only exception is the status attribute. To improve handling, we will use a simple label encoder:
(Transformation der nominalen Variable in eine binäre Variable) --> Dummys

Country ist ein Problem, da nicht numerisch. --> brauchen wir nicht für Vorhersage, daher wird der nicht umkodiert. 

In [16]:
life_expectancy.Status.value_counts()

Developing    2426
Developed      512
Name: Status, dtype: int64

In [17]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()

In [18]:
life_expectancy.Status = encoder.fit_transform(life_expectancy.Status)

Transformation vom Typ Object zum Integer

In [19]:
life_expectancy.Status.value_counts()

1    2426
0     512
Name: Status, dtype: int64

Interesting statistics of our attributes:

In [20]:
life_expectancy

Country  Year  Status  Life expectancy  Adult Mortality  \
0     Afghanistan  2015       1             65.0            263.0   
1     Afghanistan  2014       1             59.9            271.0   
2     Afghanistan  2013       1             59.9            268.0   
3     Afghanistan  2012       1             59.5            272.0   
4     Afghanistan  2011       1             59.2            275.0   
...           ...   ...     ...              ...              ...   
2933     Zimbabwe  2004       1             44.3            723.0   
2934     Zimbabwe  2003       1             44.5            715.0   
2935     Zimbabwe  2002       1             44.8             73.0   
2936     Zimbabwe  2001       1             45.3            686.0   
2937     Zimbabwe  2000       1             46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles  \
0                62     0.01               71.279624         65.0     1154   
1                64     0.01               73.523582         62.0      492   
2                66     0.01               73.219243         64.0      430   
3                69     0.01               78.184215         67.0     2787   
4                71     0.01                7.097109         68.0     3013   
...             ...      ...                     ...          ...      ...   
2933             27     4.36                0.000000         68.0       31   
2934             26     4.06                0.000000          7.0      998   
2935             25     4.43                0.000000         73.0      304   
2936             25     1.72                0.000000         76.0      529   
2937             24     1.68                0.000000         79.0     1483   

      ...  Polio  Total expenditure  Diphtheria  HIV/AIDS         GDP  \
0     ...    6.0               8.16        65.0       0.1  584.259210   
1     ...   58.0               8.18        62.0       0.1  612.696514   
2     ...   62.0               8.13        64.0       0.1  631.744976   
3     ...   67.0               8.52        67.0       0.1  669.959000   
4     ...   68.0               7.87        68.0       0.1   63.537231   
...   ...    ...                ...         ...       ...         ...   
2933  ...   67.0               7.13        65.0      33.6  454.366654   
2934  ...    7.0               6.52        68.0      36.7  453.351155   
2935  ...   73.0               6.53        71.0      39.8   57.348340   
2936  ...   76.0               6.16        75.0      42.1  548.587312   
2937  ...   78.0               7.10        78.0      43.5  547.358878   

      Population  thinness  1-19 years  thinness 5-9 years  \
0     33736494.0                  17.2                17.3   
1       327582.0                  17.5                17.5   
2     31731688.0                  17.7                17.7   
3      3696958.0                  17.9                18.0   
4      2978599.0                  18.2                18.2   
...          ...                   ...                 ...   
2933  12777511.0                   9.4                 9.4   
2934  12633897.0                   9.8                 9.9   
2935    125525.0                   1.2                 1.3   
2936  12366165.0                   1.6                 1.7   
2937  12222251.0                  11.0                11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427       10.0  
2936                            0.427        9.8  
2937                            0.434        9.8  

[2938 row

By taking a closer look at null values for each attribute we can see that there is quite a big difference in completeness:

In [ ]:
life_expectancy.isnull().sum()

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

While this information already is quite useful, we always have to put the number of missing values into relation to the absolute number of records in the dataset.

For this, we can just divide the numbers from above by the overall number of rows:



In [ ]:
life_expectancy.isnull().sum()/len(life_expectancy)

Country                            0.000000
Year                               0.000000
Status                             0.000000
Life expectancy                    0.003404
Adult Mortality                    0.003404
infant deaths                      0.000000
Alcohol                            0.066031
percentage expenditure             0.000000
Hepatitis B                        0.188223
Measles                            0.000000
 BMI                               0.011572
under-five deaths                  0.000000
Polio                              0.006467
Total expenditure                  0.076923
Diphtheria                         0.006467
 HIV/AIDS                          0.000000
GDP                                0.152485
Population                         0.221920
 thinness  1-19 years              0.011572
 thinness 5-9 years                0.011572
Income composition of resources    0.056841
Schooling                          0.055480
dtype: float64

We can also format these numbers as percentages to make them easier to interpret:

In [ ]:
(life_expectancy.isnull().sum()/len(life_expectancy)).apply('{0:.3%}'.format)

Country                             0.000%
Year                                0.000%
Status                              0.000%
Life expectancy                     0.340%
Adult Mortality                     0.340%
infant deaths                       0.000%
Alcohol                             6.603%
percentage expenditure              0.000%
Hepatitis B                        18.822%
Measles                             0.000%
 BMI                                1.157%
under-five deaths                   0.000%
Polio                               0.647%
Total expenditure                   7.692%
Diphtheria                          0.647%
 HIV/AIDS                           0.000%
GDP                                15.248%
Population                         22.192%
 thinness  1-19 years               1.157%
 thinness 5-9 years                 1.157%
Income composition of resources     5.684%
Schooling                           5.548%
dtype: object

We'll be using some algorithms that cannot deal with null values, so we will have to remove them somehow. We could use a more sophisticated approach like filling null values with the mean or median of the attributes but our dataset has a sufficient size even after dropping the rows with missing values, so it seems feasible to simply drop these rows.

In [21]:
life_expectancy = life_expectancy.dropna()

In [22]:
life_expectancy.isnull().sum()

Country                            0
Year                               0
Status                             0
Life expectancy                    0
Adult Mortality                    0
infant deaths                      0
Alcohol                            0
percentage expenditure             0
Hepatitis B                        0
Measles                            0
BMI                                0
under-five deaths                  0
Polio                              0
Total expenditure                  0
Diphtheria                         0
HIV/AIDS                           0
GDP                                0
Population                         0
thinness  1-19 years               0
thinness 5-9 years                 0
Income composition of resources    0
Schooling                          0
dtype: int64

Überprüfung passt. Es sind keine Nullwerte mehr vorhanden. 

In [23]:
len(life_expectancy)

1649

Wir haben also noch 1649 Zeilen mitz denen wir arbeiten können. 

Wir haben auch überlegt die Mittelwertimputation zu machen, aber da es noch genug Zeilen gibt mit denen wir arbeiten können, haben wir uns dazu entschieden, dies so zu lassen. 

# Data Preperation

We're now ready to select the independent and dependent variables from our dataset:

In [28]:
life_expectancy['Adult Mortality']

0       263.0
1       271.0
2       268.0
3       272.0
4       275.0
        ...  
2933    723.0
2934    715.0
2935     73.0
2936    686.0
2937    665.0
Name: Adult Mortality, Length: 1649, dtype: float64

Wir entfernen die Spalten, die wir nicht für unsere Vorhersagen benötigen. Zielvariable ist "Life Expactancy" (Y-Variable)

In [29]:
X = life_expectancy.drop(columns=['Life expectancy', 'Year', 'Country'])
y = life_expectancy['Life expectancy']

And split it into train and test sets:

Wir erstellen ein Testdatensatz, mit welchem wir unser Modell testen können. Da
für nutzen wir die train_test_split Methode und einem 80/20 Split. 

In [30]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Model Training & Evaluation

We're now ready to train some regression models on our dataset. We will test different models and compare their performance. By help of sklearn we will apply:

- Linear Regression (with and without regularization)
- Support Vector Mechine (SVM)
- Decision Tree
- Gradient Boosting